pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,995 @@
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> DispatchResult {
+		match schedule.order {
+			Order::Sell {
+				asset_in, amount_in, ..
+			} => {
+				let transaction_fee = Self::get_transaction_fee(asset_in)?;
+				ensure!(amount_in > transaction_fee, Error::<T>::TradeAmountIsLessThanFee);
+			}
+			Order::Buy { .. } => {
+				//For buy we don't check as the calculated amount in will always include the fee
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn execute_schedule(current_blocknumber: T::BlockNumber, weight: &mut u64, schedule_id: ScheduleId) {
+		*weight += Self::get_execute_schedule_weight();
+
+		let schedule = exec_or_return_if_none!(Schedules::<T>::get(schedule_id));
+		let origin: OriginFor<T> = Origin::<T>::Signed(schedule.owner.clone()).into();
+		let blocknumber_for_schedule = exec_or_return_if_none!(current_blocknumber.checked_add(&schedule.period));
+
+		let sold_currency = schedule.order.get_asset_in();
+		if exec_or_return_if_err!(Self::price_change_is_bigger_than_max_allowed(
+			sold_currency,
+			schedule.order.get_asset_out()
+		)) {
+			exec_or_return_if_err!(Self::plan_schedule_for_block(blocknumber_for_schedule, schedule_id));
+			return;
+		}
+
+		let amount_to_unreserve = exec_or_return_if_err!(Self::amount_to_unreserve(&schedule.order));
+		let remaining_amount_to_use = exec_or_return_if_none!(RemainingAmounts::<T>::get(schedule_id));
+
+		T::Currency::unreserve_named(
+			&NAMED_RESERVE_ID,
+			sold_currency.into(),
+			&schedule.owner,
+			amount_to_unreserve.into(),
+		);
+
+		exec_or_return_if_err!(Self::decrease_remaining_amount(schedule_id, amount_to_unreserve));
+
+		if remaining_amount_to_use < amount_to_unreserve {
+			Self::complete_dca(&schedule.owner, schedule_id);
+			return;
+		}
+
+		exec_or_return_if_err!(Self::take_transaction_fee_from_user(&schedule.owner, &schedule.order));
+		let trade_result = Self::execute_trade(origin, &schedule.order);
+
+		match trade_result {
+			Ok(_) => {
+				exec_or_return_if_err!(Self::plan_schedule_for_block(blocknumber_for_schedule, schedule_id));
+			}
+			_ => {
+				exec_or_return_if_err!(Self::suspend_schedule(&schedule.owner, schedule_id));
+			}
+		}
+	}
+
+	fn price_change_is_bigger_than_max_allowed(asset_a: T::Asset, asset_b: T::Asset) -> Result<bool, DispatchResult> {
+		let current_price = Self::get_current_price(asset_a, asset_b)?;
+		let price_from_short_oracle = Self::get_price_from_short_oracle(asset_a, asset_b)?;
+
+		let max_allowed = FixedU128::from(T::MaxPriceDifference::get());
+		let max_allowed_difference = current_price.saturating_mul(max_allowed);
+
+		let diff = if current_price > price_from_short_oracle {
+			current_price.saturating_sub(price_from_short_oracle)
+		} else {
+			price_from_short_oracle.saturating_sub(current_price)
+		};
+
+		Ok(diff > max_allowed_difference)
+	}
+
+	fn get_current_price(asset_a: T::Asset, asset_b: T::Asset) -> Result<FixedU128, DispatchError> {
```

Comment:
why this need to be a special function?
it is just one line. I dont think it is necessary.

why not called T::Spotpriceprovider::get_price everywhere you need instead ?